src/vit-based/vit.jl

```diff
@@ -0,0 +1,174 @@
+include("utilities.jl")
+
+# Utility function for applying LayerNorm before a block
+prenorm(planes, fn) = Chain(fn, LayerNorm(planes))
+
+"""
+      mlpblock(inplanes, hidden_planes, dropout=0.)
+  
+MLP block as used in the base ViT architecture.
+([reference](https://arxiv.org/abs/2010.11929)).
+
+# Arguments
+- `inplanes`: number of input channels
+- `hidden_planes`: number of hidden channels
+- `dropout`: dropout rate
+"""
+function mlpblock(inplanes, hidden_planes, dropout = 0.)
+  Chain(Dense(inplanes, hidden_planes, gelu),
+        Dropout(dropout),
+        Dense(hidden_planes, inplanes),
+        Dropout(dropout))
+end
+
+struct attention
+  heads
+  scale
+  qkvlayer
+  outlayer
+end
+
+"""
+    attention(planes; heads = 8, headplanes = 64, dropout = 0.)
+
+Multi-head attention block as used in the base ViT architecture.
+([reference](https://arxiv.org/abs/2010.11929)).
+
+# Arguments
+- `planes`: number of input channels
+- `heads`: number of attention heads
+- `headplanes`: number of hidden channels per head
+- `dropout`: dropout rate
+"""
+function attention(planes; heads = 8, headplanes = 64, dropout = 0.)
+  hidden_planes = headplanes * heads
+  outproject = !(heads == 1 && headplanes == planes)
+  
+  to_qkv = Dense(planes, hidden_planes * 3; bias = false)
+  to_out = outproject ? Chain(Dense(hidden_planes, planes), Dropout(dropout)) : identity
+
+  attention(heads, headplanes ^ -0.5, to_qkv, to_out)
+end
+
+function (m::attention)(x)
+  q, k, v = chunk(m.qkvlayer(x), 3; dim = 1)
+  @cast q[h, b, d, n] := q[(h, d), n, b] h in 1:m.heads
+  @cast k[h, b, d, n] := k[(h, d), n, b] h in 1:m.heads
+  @cast v[h, b, d, n] := v[(h, d), n, b] h in 1:m.heads
+  dots = batchmul(q, permutedims(k, (2, 1, 3, 4))) * m.scale
+  attn = softmax(dots; dims = 3)
+  out = batchmul(attn, v)
+  @cast out[(h, d), n, b] := out[h, b, d, n] in 1:m.heads
+  m.outlayer(out)
+end
+
+@functor attention
+
+struct Transformer
+  layers
+end
+
+"""
+    Transformer(planes, depth, heads, headplanes, mlppanes, dropout = 0.)
+
+Transformer as used in the base ViT architecture.
+([reference](https://arxiv.org/abs/2010.11929)).
+
+# Arguments
+- `planes`: number of input channels
+- `depth`: number of layers
+- `heads`: number of attention heads
+- `headplanes`: number of hidden channels per head
+- `mlppanes`: number of hidden channels in the MLP block
+- `dropout`: dropout rate
+"""
+function Transformer(planes, depth, heads, headplanes, mlpplanes, dropout = 0.)
+  layers = [Chain(SkipConnection(prenorm(planes, attention(planes; heads, headplanes, dropout)), +),
+                  SkipConnection(prenorm(planes, mlpblock(planes, mlpplanes, dropout)), +)) for _ in 1:depth]
+
+  Transformer(layers)
+end
+
+function (m::Transformer)(x)
+  for (attn, fn) in m.layers
+    x = attn(x)
+    x = fn(x)
+  end
+  return x
+end
```

Comment:
Similar comment here. This is just a `Chain`, so no need to make a new type.

Reply:
Yeah I was experimenting with a lot of things here before all the boilerplate was whittled away, so somehow that code just stayed 😂 I've trimmed it